sympy/polys/series/flint_powerseriesring.py

```diff
@@ -0,0 +1,802 @@
+                return poly
+            return fmpz_series(poly, prec=ring_prec)
+
+        with _global_cap(ring_prec):
+            return s * n
+
+    def divide(self, s1: ZZSeries, s2: ZZSeries) -> ZZSeries:
+        """Divide two power series."""
+        ring_prec = self._prec
+        if isinstance(s1, fmpz_poly):
+            s1 = fmpz_series(s1.coeffs(), prec=ring_prec)
+        if isinstance(s2, fmpz_poly):
+            s2 = fmpz_series(s2.coeffs(), prec=ring_prec)
+
+        with _global_cap(ring_prec):
+            return s1 / s2
+
+    def pow_int(self, s: ZZSeries, n: int) -> ZZSeries:
+        """Raise a power series to a non-negative integer power."""
+        ring_prec = self._prec
+        if n < 0:
+            raise ValueError("Power must be non-negative")
+
+        if isinstance(s, fmpz_poly):
+            if s.degree() * n < ring_prec:
+                return s**n
+
+            if s.degree() > ring_prec:
+                s = self.truncate(s, ring_prec)
+            poly = s**n
+            return fmpz_series(poly, prec=ring_prec)
+
+        with _global_cap(ring_prec):
+            return s**n
+
+    def square(self, s: ZZSeries) -> ZZSeries:
+        """Compute the square of a power series."""
+        return self.pow_int(s, 2)
+
+    def compose(self, s1: ZZSeries, s2: ZZSeries) -> ZZSeries:
+        """Compose two power series, `s1(s2)`."""
+        dom: Domain[MPZ] = self._domain
+        ring_prec: int = self._prec
+
+        if isinstance(s1, fmpz_poly) and isinstance(s2, fmpz_poly):
+            comp: fmpz_poly = s1(s2)
+
+            if comp.degree() < ring_prec:
+                return fmpz_poly(comp)
+            else:
+                return fmpz_series(comp.coeffs(), prec=ring_prec)
+
+        if s2 and not dom.is_zero(s2[0]):
+            raise ValueError(
+                "Series composition requires the second series to have a zero constant term."
+            )
+
+        if isinstance(s1, fmpz_poly):
+            prec2: int = _get_series_precision(s2)
+            s1 = fmpz_series(s1, prec=prec2)
+        if isinstance(s2, fmpz_poly):
+            prec1: int = _get_series_precision(s1)
+            s2 = fmpz_series(s2, prec=prec1)
+
+        with _global_cap(ring_prec):
+            return s1(s2)
+
+    def inverse(self, s: ZZSeries) -> ZZSeries:
+        """Compute the multiplicative inverse of a power series."""
+        dom: Domain[MPZ] = self._domain
+        ring_prec: int = self._prec
+        coeffs: list[MPZ] = s.coeffs()
+
+        if not coeffs or not dom.is_unit(coeffs[0]):
+            raise NotReversible(
+                "Series inverse requires the constant term to be a unit"
+            )
+
+        prec: int = (
+            _get_series_precision(s) if isinstance(s, fmpz_series) else ring_prec
+        )
+
+        dup = dup_reverse(coeffs)
+        inv = dup_truncate(dup_revert(dup, prec, dom), prec, dom)
+        return fmpz_series(inv[::-1], prec=prec)
```

Comment:
I feel like we already had this conversation but I don't remember the answer.

Why is this using `dup_revert` rather than using FLINT directly:
```python
In [3]: import flint

In [4]: x = flint.fmpz_series([0, 1])

In [5]: x
Out[5]: x + O(x^10)

In [6]: 1/(1 + x)
Out[6]: 1 + (-1)*x + x^2 + (-1)*x^3 + x^4 + (-1)*x^5 + x^6 + (-1)*x^7 + x^8 + (-1)*x^9 + O(x^10)
```